src/fivetran-destination/src/destination/dml.rs

```diff
@@ -388,6 +407,152 @@ async fn update_file(
+
+    let mut total_count = 0;
+    for path in delete_files {
+        let reader = load_file(file_config, path)
+            .await
+            .with_context(|| format!("loading delete file {path}"))?;
+        let copied = copy_delete_file(file_config, client, &insert_stmt, reader)
+            .await
+            .with_context(|| format!("handling delete file {path}"))?;
+        total_count += copied;
+    }
+    tracing::info!(total_count, "inserted rows to {schema}.{temp_table_name}");
+
+    // Skip the update if there are no rows to delete!
+    if total_count > 0 {
+        // Mark rows as deleted by "merging" our temporary table into the destination table.
+        //
+        // TODO(parkmycar): Fivetran provides a column name in truncate for soft deletes, should the
+        // delete operation be doing the same?
+
+        // HACKY: We want to update the "_fivetran_synced" column for all of the rows we marked as
+        // deleted, but don't have a way to read from the temp table that would allow this in an
+        // `UPDATE` statement.
+        let synced_time_stmt = format!(
+            "SELECT MAX({synced_col}) FROM {schema_name}.{temp_table_name}",
+            synced_col = escape::escape_identifier(FIVETRAN_SYSTEM_COLUMN_SYNCED)
+        );
+        let synced_time: SystemTime = client
+            .query_one(&synced_time_stmt, &[])
+            .await
+            .and_then(|row| row.try_get(0))?;
+
+        let matching_cols = columns.iter().filter(|col| col.is_primary);
+        let merge_stmt = format!(
+            r#"
+            UPDATE {schema_name}.{table_name}
+            SET {deleted_col} = true, {synced_col} = $1
+            WHERE ({cols}) IN (
+                SELECT {cols}
+                FROM {schema_name}.{temp_table_name}
+            )"#,
+            deleted_col = escape::escape_identifier(FIVETRAN_SYSTEM_COLUMN_DELETE),
+            synced_col = escape::escape_identifier(FIVETRAN_SYSTEM_COLUMN_SYNCED),
+            cols = matching_cols.map(|col| &col.name).join(","),
+        );
+        let total_count = client.execute(&merge_stmt, &[&synced_time]).await?;
+        tracing::info!(?total_count, "altered rows in {schema}.{table_name}");
+    }
+
+    // Cleanup, drop our temporary table.
+    client
+        .batch_execute(&format!("DROP TABLE {schema_name}.{temp_table_name}"))
+        .await?;
+
+    Ok(())
+}
+
+async fn copy_delete_file(
+    file_config: &FileConfig,
+    client: &tokio_postgres::Client,
+    insert_stmt: &tokio_postgres::Statement,
+    reader: AsyncCsvReader,
+) -> Result<usize, anyhow::Error> {
+    let mut count = 0;
+    let mut stream = reader.into_byte_records();
+    while let Some(record) = stream.try_next().await? {
+        let params = record.iter().map(|value| TextFormatter {
+            value,
+            null_string: &file_config.null_string,
+        });
+        client.execute_raw(insert_stmt, params).await?;
```

Comment:
Executing one INSERT per row is the slowest possible thing we can do. Each execution is in its own transaction, has to parse a whole statement, and wait for the network roundtrip before working on the next row.

My preference would be to just do the work to support COPY FROM HTTP/S3 because we've already written the code to do that, it just never got fully tested and merged. If that is going to take too long and this feature is more urgent, then this needs to use a single COPY FROM STDIN for the whole file/set of files and write to it. COPY FROM STDIN doesn't need to wait for any responses, so you should be able to pipeline send all them down the wire. Use the `client.copy_in[_simple]` method for that.

Reply:
Updated to use `COPY`!